RV-Monitor/Kernel_Safety_Claims_by_Runtime_Verification_Monitors.md

```diff
@@ -0,0 +1,542 @@
+* The Runtime Verification Monitor provides an architectural protection mechanisms against Kernel interferences following a safety analysis of the resources to be monitored against interference failure modes  
+
+
+### When is it valuable to use RV Monitors?
+
+From a conceptual perspective the RV Monitors introduce redundancy inside the Kernel, it is therefore valuable to use them as qualification measure when the following criteria are met:
+
+
+
+* The complexity of the Automata model is significantly lower than the code being monitored (i.e. when the code is capable of providing many functionalities under many different condition and states but only a specific subset of its potentials are used in a safety context); this implies that from a FuSa qualification effort perspective and continuous certification perspective dealing with a significantly smaller code baseline is cheaper and faster and hence it is better to make a systematic capability claim on the RV Monitor code instead of spending effort on proving the same for the monitored code.
+* The expected behavior of the monitored code is comprehensively analyzed in how it would meet the allocated safety requirements. This is needed to guarantee that the monitor is monitoring all the critical states and data for safety reasons and to avoid the monitor triggering due to valid states and data not being analyzed. 
+* The specific use case being evaluated is compatible with ex-post-facto treatment of the failure: by design the RVM will be able to detect the failure only after it has taken place and it cannot avoid it (it works as detection measure, not as prevention one) .
+* It is possible to determine that the RVM shall always trigger before the detected failure propagates in a dangerous way to the safety workloads being monitored (avoid failure propagation)
+
+
+# Integration of Runtime Verification Monitors with the Kernel
+
+A monitor is the central part of the runtime verification of a system. The monitor stands in between the formal specification of the desired (or undesired) behavior, and the trace of the actual system.
+
+In Linux terms, the runtime verification monitors are encapsulated inside the RV monitor abstraction. A RV monitor includes a reference model of the system, a set of instances of the monitor defined at development time (e.g. per-cpu monitor, per-task monitor, and so on), and the helper functions that glue the monitor to the system. Depending on both the parameters being monitored and the specific hardware features available, different approaches can provide flexibility in managing the need of monitoring vs the overhead that it might introduce. More on this later.
+
+Callbacks must be introduced in a way that is compatible with safety requirements allocated to the functionality they support (i.e. adding tracepoint by patching the code rather than by compiler could be more cost-effective from a FuSa effort point of view).
```

Comment:
... more cost-effective, but less reliable, from a FuSa ...

Reply:
because it adds one extra step which involves the kernel modifying itself.
It can have a level of reliability at most on par, but more likely lower, than the compiler time option.
I think this is beyond any debate. Self modifying code is inherently less safe, because it has a higher exposure to the risk of bugs.

Reply:
Ok I can rephrase as "(i.e. adding tracepoint by patching the code rather than by compiler could be more cost-effective from a FuSa effort point of view, but possibly more error-prone)"

Reply:
The part that really bothers me is to associate higher cost effectiveness to FuSa perspective.
FuSa is about making judgement calls that are specific to certain situations, so it is very misleading, IMO, to make such generalisation. I would leave FuSa out of this.
And even cost effectiveness, that is highly subjective. If I get sued because it turns out that I saved few developers-month by not doing the less-error prone thing. It makes I had the wrong FuSa perspective and also I was possibly not cost effective.


I think there are only 2 facts:
1.  doing runtime patching is riskier and error prone
2.  doing runtime patching here has only one potential advantage: to not invest into implementing a safer way for the RVM to connect to the rest.

I think we agree on the 1st.
WRT the second, I can only point out that this is the most objective view on it.
Then of course, if there is no budget, one tries to make ends meet, but I do not really see any advantage in retrofitting a mechanism that was designed with very different requirements in mind.

Reply:
What about "[...]. I.e. adding tracepoint by patching the code rather than by compiler would avoid the effort of qualifying the compiler mechanism, however it would be more error-prone. So for example if the number of tracepoints if very less then the direct code patching could be a viable solution, whereas for a monitor with a significant number of tracepoints the compiler tracepoints solution would be more appropriate"

I think the tradeoff above reflects the reality .....

Reply:
Then, I would just leave out these considerations.
They are still bound to an evaluation of feasibility based on arbitrary assumptions.

Reply:
Ok let's just have "Callbacks must be introduced in a way that is compatible with safety requirements allocated to the functionality they support" without further specifications

Reply:
ok